bazel/generate_objc.bzl

```diff
@@ -0,0 +1,243 @@
+    )
+
+    label_package = _join_directories([ctx.label.workspace_root, ctx.label.package])
+
+    files_with_rpc = [label_to_file(f) for f in ctx.attr.files_with_service]
+    for proto in protos:
+        outs += [
+            proto_path_to_generated_filename(
+                _GENERATED_PROTOS_DIR + "/" +
+                _get_directory_from_proto(proto) + _get_slash_from_proto(proto) +
+                to_upper_camel_with_extension(_get_file_name_from_proto(proto), "proto"),
+                _PROTO_HEADER_FMT,
+            )
+        ]
+        outs += [
+            proto_path_to_generated_filename(
+                _GENERATED_PROTOS_DIR + "/" +
+                _get_directory_from_proto(proto) + _get_slash_from_proto(proto) +
+                to_upper_camel_with_extension(_get_file_name_from_proto(proto), "proto"),
+                _PROTO_SRC_FMT,
+            )
+        ]
+
+        file_path = _strip_package_from_path(label_package, proto)
+        if not file_path.startswith("//"):
+            pass # TODO: check package boundary
+        if file_path in files_with_rpc or proto.path in files_with_rpc:
+            outs += [
+                proto_path_to_generated_filename(
+                    _GENERATED_PROTOS_DIR + "/" +
+                    _get_directory_from_proto(proto) + _get_slash_from_proto(proto) +
+                    to_upper_camel_with_extension(_get_file_name_from_proto(proto), "proto"),
+                    _GRPC_PROTO_HEADER_FMT,
+                )
+            ]
+            outs += [
+                proto_path_to_generated_filename(
+                    _GENERATED_PROTOS_DIR + "/" +
+                    _get_directory_from_proto(proto) + _get_slash_from_proto(proto) +
+                    to_upper_camel_with_extension(_get_file_name_from_proto(proto), "proto"),
+                    _GRPC_PROTO_SRC_FMT,
+                )
+            ]
+    
+    out_files = [ctx.actions.declare_file(out) for out in outs]
+    dir_out = _join_directories([
+        str(ctx.genfiles_dir.path + proto_root), label_package, _GENERATED_PROTOS_DIR
+    ])
+
+    arguments = []
+    if ctx.executable.plugin:
+        arguments += get_plugin_args(
+            ctx.executable.plugin,
+            [],
+            dir_out,
+            False,
+        )
+        tools = [ctx.executable.plugin]
+    arguments += ["--objc_out=" + dir_out]
+    arguments += get_include_protoc_args(protos)
+
+    # Include the output directory so that protoc puts the generated code in the
+    # right directory.
+    arguments += ["--proto_path={0}{1}".format(dir_out, proto_root)]
+    arguments += ["--proto_path={}".format(_get_directory_from_proto(proto)) for proto in protos]
+    arguments += [_get_srcs_file_path(proto) for proto in protos]
+
+    # create a list of well known proto files if the argument is non-None
+    well_known_proto_files = []
+    if ctx.attr.use_well_known_protos:
+        f = ctx.attr.well_known_protos.files.to_list()[0].dirname
```

Comment:
Maybe I did not understand Starlark very well, but `well_known_protos` is a boolean. Why does it have a `file` attribute?

Reply:
@muxi They are... `use_well_known_protos` (bool) and `well_known_potos`(label)